torchtune/training/_distributed.py

```diff
@@ -546,3 +550,71 @@ def shard_model(
 
     # Finally shard the entire model to account for any stragglers
     fully_shard(model, **fsdp_kwargs)
+
+
+def shard_attention_params_for_tp(
+    model: nn.Module,
+    tp_mesh: DeviceMesh,
+) -> nn.Module:
+    """
+    Shard attention parameters (heads, kv_heads, embed_dim) across tensor parallel devices.
+    Each device will handle a portion of the attention computations by having its parameters
+    scaled down by the tp device mesh size.
+
+    Args:
+        model (nn.Module): The model whose attention parameters will be sharded.
+        tp_mesh (DeviceMesh): Tensor parallelism mesh.
+
+    Returns:
+        nn.Module: Model with sharded attention parameters.
+
+    Raises:
+        ValueError: If the number of attention heads, kv heads, or embed dim is not divisible by
+        the tp device mesh size.
+
+    Example:
+        >>> # Create a model and device mesh for tensor parallelism
+        >>> model = TransformerDecoder(
+        ...     num_heads=32,
+        ...     num_kv_heads=32,
+        ...     embed_dim=4096
+        ... )
+        >>> tp_mesh = DeviceMesh("cuda", torch.arange(2))  # 2 GPUs
+        >>> # Adjust attention for tensor parallelism
+        >>> model = adjust_attention_for_tp(model, tp_mesh)
+        >>> # Now each GPU has:
+        >>> # num_heads = 16 (32/2)
+        >>> # num_kv_heads = 16 (32/2)
+        >>> # embed_dim = 2048 (4096/2)
+    """
+    # Consider the case of Deep Fusion models
+    if isinstance(model, DeepFusionModel):
+        model = model.decoder
+    tp_size = tp_mesh.size()
+    for layer in model.layers:
+        # Adjust attention module to use the local number of heads
+        attention_layers = (
+            [layer.attn]
+            if not isinstance(layer, FusionLayer)
+            else [layer.fusion_layer.attn, layer.layer.attn]
+        )
+        for attn in attention_layers:
+            if attn.num_heads % tp_size != 0:
+                raise ValueError(
+                    f"Number of attention heads ({attn.num_heads}) must be divisible by "
+                    f"tensor parallel size ({tp_size})."
+                )
+            if attn.num_kv_heads % tp_size != 0:
+                raise ValueError(
+                    f"Number of KV heads ({attn.num_kv_heads}) must be divisible by "
+                    f"tensor parallel size ({tp_size})."
+                )
+            if attn.embed_dim % tp_size != 0:
+                raise ValueError(
+                    f"Embedding dimension ({attn.embed_dim}) must be divisible by "
+                    f"tensor parallel size ({tp_size})."
+                )
+            attn.num_heads = attn.num_heads // tp_size
+            attn.num_kv_heads = attn.num_kv_heads // tp_size
+            attn.embed_dim = attn.embed_dim // tp_size
```

Comment:
Just want to make sure I understand the purpose of this utility: the idea is to set the appropriate params on the attention module so that any reshapes etc performed there will result in the correctly-shaped input for sharded Q, K, V, output projections?

Reply:
Also wonder whether we could just iterate over e.g. `model.modules()` and check `isinstance(m, MultiHeadAttentionLayer)` to avoid any dependency on the actual higher-level model arch details (fusion vs not, etc). For training this might be risky (e.g. AC will wrap modules and then you'd need to decide whether to apply on the AC-wrapped module vs not), but I don't see an obvious case it would break in inference.